tracee-ebpf/btfinfo.go

```diff
@@ -0,0 +1,126 @@
+package main
+
+import (
+	"bufio"
+	"errors"
+	"os"
+	"strings"
+	"syscall"
+)
+
+// Helpers
+
+func FileExists(path string) bool {
+	_, err := os.Stat(path)
+	return err == nil
+}
+
+func FileGrepUnique(path string, needle string) string {
+	file, _ := os.Open(path)
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+	for i := 1; scanner.Scan(); i++ {
+		if strings.Contains(scanner.Text(), needle) {
+			return scanner.Text()
+		}
+	}
+
+	return ""
+}
+
+func ByteToString(all [65]int8) string {
+	var temp [65]byte
+	for i, b := range all {
+		temp[i] = byte(b)
+	}
+	conv := string(temp[:])
+	conv = conv[:strings.Index(conv, "\x00")]
+
+	return conv
+}
+
+func KernelVersion() string {
+	utsname := syscall.Utsname{}
+	err := syscall.Uname(&utsname)
+	if err != nil {
+		return ""
+	}
+
+	return ByteToString(utsname.Release)
+}
+
+// BTFinfo
+
+type BTFinfo struct {
+	distroid     string
+	distrover    string
+	distrokernel string
+}
+
+func (btfi *BTFinfo) GetDistroId() (string, error) {
+	if btfi.distroid == "" {
+		return "", errors.New("should call discover first")
+	}
+	return btfi.distroid, nil
+}
+
+func (btfi *BTFinfo) GetDistroVer() (string, error) {
+	if btfi.distrover == "" {
+		return "", errors.New("should call discover first")
+	}
+	return btfi.distrover, nil
+}
+
+func (btfi *BTFinfo) GetDistroKernel() (string, error) {
+	if btfi.distrokernel == "" {
+		return "", errors.New("should call discover first")
+	}
+	return btfi.distrokernel, nil
+}
+
+func (btfi *BTFinfo) DiscoverDistroByKernel() error {
+	// tries to discover running distribution by kernel version/name
+
+	kernel := KernelVersion()
+	// fedora:      4.18.16-300.fc29
+	// centos:      3.10.0-1160.31.1.el7.centos.x86_64
+	if strings.Contains(kernel, "fc") {
+		btfi.distroid = "fedora"
+		btfi.distrover = kernel[strings.Index(kernel, "fc")+2:]
+		btfi.distrokernel = kernel[:strings.Index(kernel, ".fc")]
+		return nil
+	} else if strings.Contains(kernel, "centos") {
+		btfi.distroid = "centos"
+		t := kernel
+		t = t[strings.Index(t, "el")+2:]
+		t = t[:strings.Index(t, ".")]
+		btfi.distrover = t
+		btfi.distrokernel = kernel[:strings.Index(kernel, ".el")]
+
+		return nil
+	} // else if ...
+
+	return errors.New("can't discover distribution by kernel version")
+}
+
+func (btfi *BTFinfo) DiscoverDistroByRelease() error {
```

Comment:
Instead of creating different discover methods, I would create just one Discover method, and try different methods until a match is found.
This function can have the following signature:
`func Discover () (BTFInfo, error)`
Then, if err=nil, the user can just extract the data from (the now exported) BTFInfo fields:
btfi.DistroVer, btfi.DistroID, btfi.DistroKernel.
WDYT?

Reply:
Yes, I think it's ok doing that now. This was a subset of a major change from: 

https://github.com/rafaeldtinoco/tracee/tree/external-btf 

which contained an Object and more methods (to get and extract BTF files from the HUB). Intention was to isolate the BTFinfo logic, and then in near future start isolating things in a more objectification approach for existing code. I'm ok in having this as a single function now and changing later if needed (or keeping as is).

Reply:
Actually, if you don't mind, I think I'll keep this as is (as an object). Just because my parallel logic (from the external-btf branch) will use it as an object for now and I dont want to re-write it. Let me know if you have hard preference on this and I can change it.

Reply:
> Instead of creating different discover methods, I would create just one Discover method, and try different methods until a match is found.

I have done that though. Just 1 discovery method.. I believe this is better also.

Reply:
```go
// NewBTFInfo creates a BTFInfo object and runs DiscoverDistro()
// on its creation. It returns nil if no discovery data was
// avalable. To get specific error from DiscoverDistro method,
// user should create a BTFInfo struct and call DiscoverDistro
// by itself.
func NewBTFInfo() *BTFInfo {
	btfi := new(BTFInfo)
	if err := btfi.DiscoverDistro(); err != nil {
		return nil
	}
	return btfi
}
```

So using it might be as simple as:

```go
	btfi := NewBTFInfo()
	if btfi == nil {
		return errors.New("BTFInfo failed")
	}
	fmt.Println(btfi.GetDistroId()
```

Reply:
Marking as resolved with my last push, please let me know if you prefer it changed. I'll expand the object with logic from external-btf branch in my repository. Thanks!